### PR TITLE
issue: 1055916 add support for UMR in MP-RQ 

### PR DESCRIPTION
**Pull request by rafiw**
only last commit is relavent

### CODE REVIEW

**Comment by rafiw** [comment]
@rosenbaumalex

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1900/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1901/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1338/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1339/ for details (Mellanox internal link).

**Comment by rafiw** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1341/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1903/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1342/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1904/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1351/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1914/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1919/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1356/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1421/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1422/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1423/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1424/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1439/ for details (Mellanox internal link).

**Comment by rafiw** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1440/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1444/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1478/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1494/ for details (Mellanox internal link).

File: src/vma/dev/allocator.h [hunk 1]
```diff
@@ -46,7 +46,11 @@ class vma_allocator {
 	~vma_allocator();
 	void* alloc_and_reg_mr(size_t size, ib_ctx_handler *p_ib_ctx_h);
 	void* get_ptr() const {return m_data_block;}
-	uint32_t find_lkey_by_ib_ctx(ib_ctx_handler *p_ib_ctx_h) const;
+	inline uint32_t find_lkey_by_ib_ctx(ib_ctx_handler *p_ib_ctx_h) const
+	{
+		return find_ibv_mr_by_ib_ctx(p_ib_ctx_h)->lkey;
```
Comment (DanielLibenson) [inline src/vma/dev/allocator.h]: Please add a check for NULL value.

File: src/vma/dev/allocator.h [hunk 2]
```diff
@@ -46,7 +46,11 @@ class vma_allocator {
 	~vma_allocator();
 	void* alloc_and_reg_mr(size_t size, ib_ctx_handler *p_ib_ctx_h);
 	void* get_ptr() const {return m_data_block;}
-	uint32_t find_lkey_by_ib_ctx(ib_ctx_handler *p_ib_ctx_h) const;
+	inline uint32_t find_lkey_by_ib_ctx(ib_ctx_handler *p_ib_ctx_h) const
+	{
```
Comment (DanielLibenson) [inline src/vma/dev/allocator.h]: Move open { bracket to the end of function declaration line.

File: src/vma/dev/ib_ctx_handler.cpp [hunk 3]
```diff
@@ -245,3 +260,167 @@ void ib_ctx_handler::handle_event_DEVICE_FATAL()
...
+
+bool ib_ctx_handler::post_umr_wr(struct ibv_exp_send_wr &wr)
+{
+#ifdef HAVE_MP_RQ
+	auto_unlocker lock(m_lock_umr);
+	ibv_exp_send_wr *bad_wr = NULL;
+	ibv_exp_wc wc;
+	int ne = 0;
+
+	if (!m_umr_qp) {
```
Comment (DanielLibenson) [inline src/vma/dev/ib_ctx_handler.cpp]: Please merge both if's statements to one if

File: src/vma/dev/ib_ctx_handler.cpp [hunk 4]
```diff
@@ -245,3 +260,167 @@ void ib_ctx_handler::handle_event_DEVICE_FATAL()
...
 	g_p_event_handler_manager->unregister_ibverbs_event(m_p_ibv_context->async_fd, this);
 }
+
+bool ib_ctx_handler::post_umr_wr(struct ibv_exp_send_wr &wr)
+{
+#ifdef HAVE_MP_RQ
+	auto_unlocker lock(m_lock_umr);
+	ibv_exp_send_wr *bad_wr = NULL;
+	ibv_exp_wc wc;
+	int ne = 0;
```
Comment (DanielLibenson) [inline src/vma/dev/ib_ctx_handler.cpp]: Declare **ne** parameter near the line it used.

File: src/vma/dev/ib_ctx_handler.cpp [hunk 5]
```diff
@@ -245,3 +260,167 @@ void ib_ctx_handler::handle_event_DEVICE_FATAL()
...
+				    "addr=%#x, length=%d, lkey=%#x",
+				    bad_wr->wr_id,
+				    bad_wr->exp_send_flags,
+				    bad_wr->sg_list[0].addr,
+				    bad_wr->sg_list[0].length,
+				    bad_wr->sg_list[0].lkey);
+		}
+		return false;
+	}
+	while (!ne) {
```
Comment (DanielLibenson) [inline src/vma/dev/ib_ctx_handler.cpp]: You may use do-while

File: src/vma/dev/ring_eth_cb.cpp [hunk 6]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+	prev_addr = base_ptr;
+	mr = m_alloc.find_ibv_mr_by_ib_ctx(m_p_ib_ctx);
+
+	if (get_partition()) {
+		net_len = ETH_VLAN_HDR_LEN + sizeof(struct iphdr) + sizeof(struct udphdr);
+	}
+	else {
+		net_len = ETH_HDR_LEN + sizeof(struct iphdr) + sizeof(struct udphdr);
+	}
+	m_payload_len = m_cb_ring.stride_bytes;
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Add separative line

File: src/vma/dev/ring_eth_cb.cpp [hunk 7]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+	m_hdr_len = m_cb_ring.hdr_bytes;
+	// in case stride smaller then packet size
+	while ((m_stride_size * count) < m_payload_len) {
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Shouldn't it be <= ?

File: src/vma/dev/ring_eth_cb.cpp [hunk 8]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+		case STRIP_NETWORK_HRDS:
+			// network not accessible to application
+			m_p_mem_rep_list[index].base_addr = base_ptr;
+			m_p_mem_rep_list[index].byte_count[0] = net_len;
+			m_p_mem_rep_list[index].stride[0] = net_len;
+			m_p_mem_rep_list[index].mr = mr;
+			curr_data_len = packets_num * net_len;
+			prev_addr += curr_data_len;
+			index++;
+			if (m_hdr_len) {
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Add separative line

File: src/vma/dev/ring_eth_cb.cpp [hunk 9]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+			index++;
+		break;
+		case SEPERATE_NETWORK_HRDS:
+			if (m_hdr_len) {
+				packet_len += m_hdr_len;
+				// for size calculation in read_cyclic
+				m_hdr_len = packet_len;
+			} else {
+				m_hdr_len = net_len;
+			}
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Add separative line

File: src/vma/dev/ring_eth_cb.cpp [hunk 10]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+	m_p_mem_rep_list[index].stride[0] = pad_len;
+	m_p_mem_rep_list[index].mr = mr;
+
+	memset(&mrin, 0, sizeof(mrin));
+	mrin.pd = m_p_ib_ctx->get_ibv_pd();
+	mrin.attr.create_flags = IBV_EXP_MR_INDIRECT_KLMS;
+	mrin.attr.exp_access_flags = IBV_EXP_ACCESS_LOCAL_WRITE;
+	mrin.attr.max_klm_list_size = m_umr_blocks;
+	m_p_umr_mr = ibv_exp_create_mr(&mrin);
+	if (!m_p_umr_mr) {
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Add separative line

File: src/vma/dev/ring_eth_cb.cpp [hunk 11]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+		ring_logdbg("Failed creating mr %m", errno);
+		goto cleanup;
+	}
+	memset(&m_umr_wr, 0, sizeof(m_umr_wr));
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Add separative line

File: src/vma/dev/ring_eth_cb.cpp [hunk 12]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+	m_umr_wr.ext_op.umr.base_addr = (uint64_t)mr->addr;
+	m_umr_wr.ext_op.umr.num_mrs = m_umr_blocks;
+	m_umr_wr.exp_send_flags |= IBV_EXP_SEND_SIGNALED;
+	m_umr_wr.exp_opcode = IBV_EXP_WR_UMR_FILL;
+	res = m_p_ib_ctx->post_umr_wr(m_umr_wr);
+	if (!res) {
+		ring_logdbg("Failed in ibv_exp_post_send IBV_EXP_WR_UMR_FILL\n");
+		goto cleanup;
+	}
+	// redmine.mellanox.com/issues/396761/
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Add separative line

File: src/vma/dev/ring_eth_cb.cpp [hunk 13]
```diff
@@ -145,7 +136,234 @@ qp_mgr* ring_eth_cb::create_qp_mgr(const ib_ctx_handler *ib_ctx,
...
+	m_umr_wr.ext_op.umr.mem_list.rb.repeat_count = m_p_rpt_cnt;
+	m_umr_wr.exp_send_flags = IBV_EXP_SEND_INLINE;
+	m_umr_wr.ext_op.umr.exp_access = IBV_EXP_ACCESS_LOCAL_WRITE;
+	m_umr_wr.ext_op.umr.modified_mr = m_p_umr_mr;
+	m_umr_wr.ext_op.umr.base_addr = (uint64_t)mr->addr;
+	m_umr_wr.ext_op.umr.num_mrs = m_umr_blocks;
+	m_umr_wr.exp_send_flags |= IBV_EXP_SEND_SIGNALED;
+	m_umr_wr.exp_opcode = IBV_EXP_WR_UMR_FILL;
+	res = m_p_ib_ctx->post_umr_wr(m_umr_wr);
+	if (!res) {
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: move **m_p_ib_ctx->post_umr_wr(m_umr_wr)** function call inside the if

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1526/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1527/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1529/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1532/ for details (Mellanox internal link).

**Comment by rafiw** [comment]
bot:retest

File: src/vma/vma_extra.h [hunk 14]
```diff
@@ -263,19 +265,35 @@ typedef enum {
...
+ * The number of headers\packets is equal to packets in @ref vma_completion_cb_t.
+ * the packet memory layout has five options:
+ * 1. RAW_PACKET - payload_ptr will point to the raw packet containing the
+ *     network headers and user payload.
+ * 2. STRIP_NETWORK_HDRS - network headers will be ignored by VMA.
+ *      payload_ptr - will point to the first packet which it size is defined in
+ *          stride_bytes.
+ *      a. hdr_bytes > 0 - usr_hdr_ptr will point to the first header.
+ *      b. hdr_bytes = 0 - usr_hdr_ptr is NULL
+ * 2. SEPERATE_NETWORK_HDRS - network headers will be dropped
```
Comment (rafiw) [inline src/vma/vma_extra.h]: network headers will be seperated

File: src/vma/dev/ring_eth_cb.cpp [hunk 15]
```diff
@@ -233,6 +224,20 @@ int ring_eth_cb::allocate_umr_mem()
 	}
 
 	pad_len = (m_stride_size * count) - net_len - m_hdr_len - m_payload_len;
+	// no need to allocate all padding add one so starting address will be valid
+	pack_written_size = (m_stride_size * count) - pad_len + 1;
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Consider to remove +1.

File: src/vma/dev/ring_eth_cb.cpp [hunk 16]
```diff
@@ -233,6 +224,20 @@ int ring_eth_cb::allocate_umr_mem()
...
+	// allocate buffer
+	buffer_size = pack_written_size * packets_num;
+	base_ptr = (uint64_t)m_alloc.alloc_and_reg_mr(buffer_size, m_p_ib_ctx);
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Check that allocation failure is treated well at 237 line.

File: src/vma/dev/ring_eth_cb.cpp [hunk 17]
```diff
@@ -252,7 +257,8 @@ int ring_eth_cb::allocate_umr_mem()
 			// network not accessible to application
 			m_p_mem_rep_list[index].base_addr = base_ptr;
 			m_p_mem_rep_list[index].byte_count[0] = net_len;
-			m_p_mem_rep_list[index].stride[0] = net_len;
+			// optimize write header to the same place
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Consider to change comment to "optimize write header to the same physical address"

File: src/vma/dev/ring_eth_cb.cpp [hunk 18]
```diff
@@ -233,6 +224,20 @@ int ring_eth_cb::allocate_umr_mem()
 	}
 
 	pad_len = (m_stride_size * count) - net_len - m_hdr_len - m_payload_len;
```
Comment (DanielLibenson) [inline src/vma/dev/ring_eth_cb.cpp]: Consider to set padding address to data address (It should help with debugging/wrong expected packet size)

**Comment by liranoz12** [comment]
bot:retest

**Comment by rafiw** [comment]
bot:retest

**Comment by liranoz12** [comment]
bot:retest

**Comment by liranoz12** [comment]
bot:retest

**Pull request merged**